### PR TITLE
Make stream-related iterators safe by default

### LINKED ISSUE #54: Iterators aren't safe yet
The goal is for HILTI's iterators to be safe in the sense that accesses to expired content gets caught. This works currently already for containers that go out of scope. For example, for a vector `v`, `begin(v)` returns an iterator that will become invalid once `v` gets destroyed. However, for mutable containers, we don't catch the case yet that changes to the container's content may invalidate iterators due to internal data structure updates. An access to such an iterator needs to throw an exception, under a general policy of "container modifications invalidate existing iterators".

### PR DESCRIPTION
**Pull request by bbannier**
This closes #54.

### CODE REVIEW

File: hilti/include/rt/types/stream.h [hunk 1]
```diff
@@ -683,18 +690,19 @@ class View {
...
         return detail::UnsafeConstIterator(_begin);
     }
 
-    detail::UnsafeConstIterator end() const { return detail::UnsafeConstIterator(safeEnd()); }
+    detail::UnsafeConstIterator unsafeEnd() const { return detail::UnsafeConstIterator(safeEnd()); }
+
+    auto begin() const { return safeBegin(); }
+    auto end() const { return safeEnd(); }
 
     const SafeConstIterator& safeBegin() const { return _begin; }
```
Comment (rsmmr) [inline hilti/include/rt/types/stream.h:703]: Do we want just want to remove these and just use begin/end for the safe versions?

File: hilti/include/rt/types/stream.h [hunk 2]
```diff
@@ -565,15 +572,15 @@ class View {
     bool isOpenEnded() const { return ! _end.has_value(); }
```
Comment (rsmmr) [inline hilti/include/rt/types/stream.h:572]: How about renaming this to `isExpandable()` now, that seems more intuitive.

File: hilti/src/rt/tests/stream.cc [hunk 3]
```diff
@@ -16,169 +16,471 @@ using hilti::rt::to_string;
...
+
+    SUBCASE("difference") {
+        const auto [s, before_begin] = []() {
+            auto s = Stream(" 123"_b);
+            const auto before_begin = s.begin();
+
+            s.trim(before_begin + 1);
+            REQUIRE_EQ(s, "123"_b);
+
+            return std::make_tuple(s, before_begin);
```
Comment (rsmmr) [inline hilti/src/rt/tests/stream.cc:348]: Isn''t `before_begin` just becoming an expired iterator here because `s` won't survive the return?

Comment (bbannier) [reply hilti/src/rt/tests/stream.cc:348]: Good catch! I added a `REQUIRE_FALSE(before_begin.isExpired())` which failed with the code in this form, but succeeds if we `std::move(s)` into the returned tuple.

File: hilti/include/rt/types/stream.h (hunk 2 above)
Comment (bbannier) [reply hilti/include/rt/types/stream.h:572]: Hmm, personally `isOpenEnded` seems clearer to me as it a pretty clear property of some range/window to me. `isExpandable` sounds like one could do something with the `View` to make it expandable, but it is really a fixed property.

File: hilti/src/rt/tests/stream.cc (hunk 3 above)
Comment (rsmmr) [reply hilti/src/rt/tests/stream.cc:348]: ok, nevermind then

File: hilti/include/rt/types/stream.h [hunk 4]
```diff
@@ -223,14 +217,18 @@ class SafeConstIterator {
...
+                e = {_content, c->offset() + c->size(), _chunk};
+            else {
+                auto l = c->last();
+                assert(l);
+                assert(l->isLast());
+                e = {_content, l->offset() + l->size(), l};
+            }
         }
+        else
+            cannot_be_reached();
```
[the remark this answers is not in the document]
Comment (bbannier) [reply hilti/include/rt/types/stream.h:228]: As far as I can tell, these truely cannot be reached (checked by inspection and confirmed by coverage in the surrounding code). I would suggest we leave this as is, and revisit should we ever run into an issue. Previously these cases potentially triggered segfaults, but now they should generate much easier to understand aborts from `cannot_be_reached`.

Resolving for now.

**Comment by bbannier** [comment]
> I would suggest one additional set of tests that exercises methods with a more complex stream instance containing multiple chunks, incl. some with SBO. You could add a helper function that creates such an instance and then reuse that across the tests to see if the various operations produce the expected result on it it as well.

I did that and augmented a number of tests with checks of multichunk streams. What does SBO refer to?

**Pull request merged**